---
title: "Create datasets for machine learning"
author: 
  - name: Levi Waldron
    affiliation:
    - CUNY Graduate School of Public Health and Health Policy
    email: levi.waldron@sph.cuny.edu
output: 
  BiocStyle::html_document:
    self_contained: yes
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: show
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('curatedMetagenomicAnalyses')`"
vignette: >
  %\VignetteIndexEntry{ML datasets from cMD}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This vignette identifies and writes to disk datasets for use in multi-dataset machine learning. For each relevant study condition, two files are written:

1. A `.rda` file containing a TreeSummarizedExperiment with full colData, rowData, a phylogenetic tree, and assay data containing taxonomic data.
2. A `.csv.gz` file containing only key sample metadata, and taxonomic data (variables in columns)

See [source file on GitHub](https://github.com/waldronlab/curatedMetagenomicAnalyses/blob/main/vignettes/MLdatasets.Rmd)

Do `BiocManager::install("waldronlab/curatedMetagenomicAnalyses")` to install the
`makeSEforCondition` function and to run this vignette.

Packages used here:

```{r, message=FALSE}
library(curatedMetagenomicData)
library(curatedMetagenomicAnalyses)
library(dplyr)
library(tidyverse)
library(glue)
```


```{r}
sampleMetadata |>
    filter(age >= 18) |>
    filter(!is.na(alcohol)) |>
    filter(body_site == "stool") |>
    select(where(~ !all(is.na(.x)))) |>
    returnSamples("relative_abundance", rownames = "short")
```

```{r}
alcoholStudy <-
    filter(sampleMetadata, age >= 18) |>
    filter(!is.na(alcohol)) |>
    filter(body_site == "stool") |>
    select(where(~ !all(is.na(.x)))) |>
    returnSamples("relative_abundance", rownames = "short")
```


# Investigate potential response variables

These are the 10 study conditions most commonly found in curatedMetagenomicData:

```{r}
data("sampleMetadata")
availablediseases <- pull(sampleMetadata, study_condition) %>%
  table() %>%
  sort(decreasing = TRUE)
availablediseases
```

And the number of studies they are found in:
```{r}
studies <- lapply(names(availablediseases), function(x){
  filter(sampleMetadata, study_condition %in% x) %>%
    pull(study_name) %>%
    unique()
})
names(studies) <- names(availablediseases)
studies <- studies[-grep("control", names(studies))] #get rid of controls
studies <- studies[sapply(studies, length) > 1] #available in more than one study
studies
```

Each of these datasets has six data types associated with it; for example:
```{r}
curatedMetagenomicData("JieZ_2017.+")
```



# Write relative abundance datasets to disk

```{r writefiles, message=FALSE, cache=TRUE}
# Unuseful block


# Pick CRC data
# cond <- names(studies)[3]
# se <-
#   curatedMetagenomicAnalyses::makeSEforCondition(cond, removestudies = c("HMP_2019_ibdmdb", "ThomasAM_2019_c", "VogtmannE_2016", "WirbelJ_2018",   "YachidaS_2019", "YuJ_2015", "ZellerG_2014"), dataType = "relative_abundance")
# print(paste("Next study condition:", cond, " /// Body site: ", unique(colData(se)$body_site)))
# print(with(colData(se), table(study_name, study_condition)))
# cat("\n \n")
# save(se, file = paste0(cond, ".rda"))
# flattext <- select(as.data.frame(colData(se)), c("study_name", "study_condition", "subject_id"))
# rownames(flattext) <- colData(se)$sample_id
# flattext <- cbind(flattext, data.frame(t(assay(se))))
# write.csv(flattext, file = paste0(cond, ".csv"))
```

```{r}
# Unuseful block

# for (i in seq_along(studies)){
#   cond <- names(studies)[i]
#   se <-
#     curatedMetagenomicAnalyses::makeSEforCondition(cond, removestudies = "HMP_2019_ibdmdb", dataType = "relative_abundance")
#   print(paste("Next study condition:", cond, " /// Body site: ", unique(colData(se)$body_site)))
#   print(with(colData(se), table(study_name, study_condition)))
#   cat("\n \n")
#   save(se, file = paste0(cond, ".rda"))
#   flattext <- select(as.data.frame(colData(se)), c("study_name", "study_condition", "subject_id"))
#   rownames(flattext) <- colData(se)$sample_id
#   flattext <- cbind(flattext, data.frame(t(assay(se))))
#   write.csv(flattext, file = paste0(cond, ".csv"))
#   system(paste0("gzip ", cond, ".csv"))
# }
```

# Direct link to files

Download the .csv and .rda files directly from https://www.dropbox.com/sh/0t0nbhj9eqm3wkq/AACZIw42WA-uHjzo97bG5tE6a?dl=0

```{r}
# get list of studies to read in
study_names <- c("FengQ_2015", "GuptaA_2019", "HanniganGD_2017", "ThomasAM_2018a", "ThomasAM_2018b", "ThomasAM_2019_c", "VogtmannE_2016", "WirbelJ_2018", "YachidaS_2019", "YuJ_2015", "ZellerG_2014" )
# map over the study names and download into SE objects
study_data <- map(study_names, ~ curatedMetagenomicData(glue("{.}.relative_abundance"), counts = TRUE, dryrun = FALSE))
```

```{r}
# unify into a single SE object
unified_data <- map(study_data, ~ assay(.[[1]])) |>
  map(~ data.frame(.) |> rownames_to_column("taxon")) |>
  purrr::reduce(full_join)

dim(unified_data)
# map(study_data, ~ dim(assay(.[[1]])))
```

```{r}
# metadata: study names and health condition
metadata <- map_dfr(
  study_data, ~ colData(.[[1]]) |>
    as_tibble() |>
    select(study_name, study_condition)
)

obs_names <- (colnames(unified_data))
obs_names <- obs_names[2:length(obs_names)]
rownames(metadata) <- obs_names
metadata <- cbind(obs_names, metadata)
colnames(metadata)[1] = "observations"

write_csv(metadata, "crc_meta.csv")
```

```{r}
numobs_list = map(study_data, ~ dim(assay(.[[1]]))[2])
study_names_col = list()
for (i in length(study_names)){
  study_names_col = list(study_names_col, rep(study_names[i], numobs_list[[i]]))
}
```

```{r}
# crc_separate_taxon.csv: rows: obs, cols: species, with separate taxon level
unified_data2 <- unified_data |>
  tidyr::separate("taxon", c("kingdom", "phylum", "class", "order", "family", "genus", "species"), sep = "\\|")

# save separate feature data: rows: obs, cols: species, with separate taxon level
unified_data2 <- t(unified_data2)
meta_temp <- data.frame(matrix(NA, nrow = 7, ncol = 3))
meta_temp[,1] <- c("kingdom", "phylum", "class", "order", "family", "genus", "species")
names(meta_temp) <- colnames(metadata)
meta_temp <- rbind(meta_temp, metadata)
unified_data2 <- cbind(meta_temp, unified_data2)
unified_data2 <- unified_data2[,2:ncol(unified_data2)]

  
write.csv(unified_data2, "crc_separate_taxon.csv")
```

```{r}
# crc_all: rows: obs, cols: study name, condition, species (whole)
unified_data3 <- t(unified_data)
colnames(unified_data3) = unified_data3[1,]
unified_data3 = unified_data3[-1,]

unified_data3 <- cbind(meta_temp, unified_data3)
colnames(unified_data3)[1] = "observations"
write_csv(unified_data3, "crc_whole_taxon.csv")
```

```{r}
unique(unified_data3[,2])
```


